proto/backend.proto

```diff
@@ -0,0 +1,217 @@
+//-----------------------------------------------
+
+message PluginStatusRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  enum FetchInfo {
+    STATUS = 0;  // nothing
+    API = 1;     // OpenAPI description
+    METRICS = 2; // Prometheus style metrics
+    DEBUG = 3;   // JSON Debug Info (admin+ view)
+  }
+
+  // Get an OpenAPI description (viewer+)
+  FetchInfo fetch = 3;
+}
+
+message PluginStatusResponse {
+  enum PluginStatus {
+    UNKNOWN = 0; // starting etc
+    OK = 1;
+    ERROR = 2;
+  }
+
+  PluginStatus status = 1;
+  string info = 2;
+}
+
+
+//-----------------------------------------------
+// QUERY
+//-----------------------------------------------
+
+message TimeRange {
+  string fromRaw = 1;
+  string toRaw = 2;
+  int64 fromEpochMs = 3;
+  int64 toEpochMs = 4;
+}
+
+message DataQuery {
+  string refId = 1;
+  int64 maxDataPoints = 2;
+  int64 intervalMs = 3;
+  TimeRange timeRange = 4; // NOTE: this moved from QueryRequest in v1
+  string json = 5; // Arbitrary structure (model?  modelJson in v1)
+}
+
+message QueryRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // List of queries
+  repeated DataQuery queries = 3;
+}
+
+message QueryResponse {
+  // Arrow encoded DataFrames
+  // Each frame encodes its own: Errors, meta, and refId
+  repeated bytes data = 1;
+
+  // Additional response metadata
+  map<string,string> headers = 2;
+}
+
+
+//-----------------------------------------------
+// HTTP style request
+//-----------------------------------------------
+
+message PluginRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // HTTP Style parameters
+  string method = 3; // GET/POST/DELETE/etc
+  string path = 4;   // (with params?)
+  bytes body = 5;    // assume string unless header says otherwise?
+}
+
+message PluginResponse {
+  int32 code = 1;  // Match HTTP status codes
+  bytes body = 2;
+  map<string,string> headers = 3;
+}
+
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Request(PluginRequest) returns (PluginResponse);
+
+  // Well typed query interface
+  rpc Query(QueryRequest) returns (QueryResponse);
+
+  // TODO -- currently queries die when the response >4mb...
+  // We need a better solution...
+  // a) encode paging info in the DataFrame metadata?
+  // b) chunk the results and grafana backend would
+}
+
+
+//-----------------------------------------------
+// Allow plugins to access other plugins
+// Only allow this for some plugins?  GEL only
+//
+// ???? How does the backend know the PluginConfig????
```

Comment:
Which PluginConfig? Not sure I understand the question.

Reply:
PluginRequest/QueryRequest both have a parameter for PluginConfig -- that includes the full datasource metadata.  From a backend plugin, how does it know this?  Maybe it just sets the ID, and the server will fill up the rest?

Reply:
In terms of GEL (Access other datasources) that is how it currently works. GEL just gets the query with the datasource ID, and passes that request on to Grafana.